bcipy/tasks/rsvp/copy_phrase.py

```diff
@@ -44,368 +71,610 @@ class RSVPCopyPhraseTask(Task):
-        self.notch_filter_frequency = self.parameters['notch_filter_frequency']
-
-        self.min_num_inq = parameters['min_inq_len']
-        self.collection_window_len = parameters['trial_length']
-
-        self.static_offset = parameters['static_trigger_offset']
-
-        # Show selection feedback
-        self.show_feedback = parameters['show_feedback']
-        self.feedback_color = parameters['feedback_message_color']
-
-        if self.show_feedback:
-            self.feedback = VisualFeedback(
-                self.window, self.parameters, self.experiment_clock)
 
-        # Preview inquiry parameters
-        self.preview_inquiry = parameters['show_preview_inquiry']
-
-    def execute(self):
-        self.logger.debug('Starting Copy Phrase Task!')
-
-        # already correctly spelled letters
-        text_task = str(self.copy_phrase[0:self.spelled_letters_count])
-        task_list = [(str(self.copy_phrase),
-                      str(self.copy_phrase[0:self.spelled_letters_count]))]
-
-        # Try Initializing Copy Phrase Wrapper:
-        copy_phrase_task = CopyPhraseWrapper(
-            self.min_num_inq,
-            self.max_inq_per_trial,
+        # TODO: add a parameter for button_press_error_prob.
+        self.button_press_error_prob = 0.05
+        self.evidence_names = [EvidenceName.LM, EvidenceName.ERP]
+        if self.parameters['show_preview_inquiry']:
+            self.evidence_names.append(EvidenceName.BTN)
+
+    def setup(self) -> None:
+        """Initialize/reset parameters used in the execute run loop."""
+
+        self.spelled_text = str(
+            self.copy_phrase[0:self.starting_spelled_letters()])
+        self.last_selection = ''
+        self.inq_counter = 0
+        self.session = Session(save_location=self.file_save,
+                               task='Copy Phrase',
+                               mode='RSVP')
+        self.write_session_data()
+
+        self.init_copy_phrase_task()
+        self.current_inquiry = self.next_inquiry()
+
+    def starting_spelled_letters(self) -> int:
+        """Number of letters already spelled at the start of the task."""
+        spelled_letters_count = self.parameters['spelled_letters_count']
+        if spelled_letters_count > len(self.copy_phrase):
+            self.logger.debug('Already spelled letters exceeds phrase length.')
+            spelled_letters_count = 0
+        return spelled_letters_count
+
+    def next_inquiry(self) -> InquirySchedule:
+        """Generate an InquirySchedule for spelling the next letter."""
+        if self.copy_phrase_task:
+            _, inquiry_schedule = self.copy_phrase_task.initialize_series()
+            return inquiry_schedule
+        return None
+
+    def init_copy_phrase_task(self) -> CopyPhraseWrapper:
+        """Initialize the CopyPhraseWrapper.
+
+        Returns:
+        --------
+        initialized CopyPhraseWrapper
+        """
+
+        self.copy_phrase_task = _init_copy_phrase_wrapper(
```

Comment:
At this point, since so many parameters are extracted and passed in, is it easier to just pass around the `parameters` dict and functions can grab contents directly?

If this adds concerns about parameters being accidentally changed inside functions, it could be changed to some immutable data structure. (For example, maybe something like: https://pypi.org/project/frozendict/. Also dataclasses with `frozen=True`  would probably be suitable for python3.7+)

Reply:
Agreed. We're inconsistent in the code as to when we pass all parameters vs when we pass parameter values as arguments to other objects. For this first pass at least I wanted to be clear when the argument value was just a straight parameter value vs a computed value. I think we should take a second pass at this in another ticket where we decide whether or not to pass the whole parameter dict around and let the consuming object take what it wants, or we have another way of organizing these similar to what Tab did regarding the display code using named tuples for a subset of parameters. 

The parameters object looks like a dict but it's actually a subclass (defined in parameters.json), so we can easily enforce immutability there. We have been treating it as an immutable object but haven't enforced that.